propulate/propagators/pso/basic_pso.py

```diff
@@ -0,0 +1,125 @@
+    Else, it might happen that warnings occur.
+
+    This propagator also serves as the foundation of all other pso propagators and supplies
+    them with protected methods that help in the update process.
+
+    If you want to implement further pso propagators, please do your best to
+    derive them from this propagator or from one that is derived from this.
+
+    This propagator works on Particle-class objects.
+    """
+
+    def __init__(
+        self,
+        w_k: float,
+        c_cognitive: float,
+        c_social: float,
+        rank: int,
+        limits: Dict[str, Tuple[float, float]],
+        rng: Random,
+    ):
+        """
+        Class constructor.
+        :param w_k: The learning rate ... somehow
+        :param c_cognitive: constant cognitive factor to scale p_best with
+        :param c_social: constant social factor to scale g_best with
+        :param rank: the rank of the worker the propagator is living on in MPI.COMM_WORLD
+        :param limits: a dict with str keys and 2-tuples of floats associated to each of them
+        :param rng: random number generator
+        """
+        super().__init__(parents=-1, offspring=1)
+        self.c_social = c_social
+        self.c_cognitive = c_cognitive
+        self.w_k = w_k
+        self.rank = rank
+        self.limits = limits
+        self.rng = rng
+        self.laa: np.ndarray = np.array(
+            list(limits.values())
+        ).T  # laa - "limits as array"
+
+    def __call__(self, particles: List[Particle]) -> Particle:
+        old_p, p_best, g_best = self._prepare_data(particles)
+
+        new_velocity: np.ndarray = (
+            self.w_k * old_p.velocity
+            + self.rng.uniform(0, self.c_cognitive) * (p_best.position - old_p.position)
+            + self.rng.uniform(0, self.c_social) * (g_best.position - old_p.position)
+        )
+        new_position: np.ndarray = old_p.position + new_velocity
+
+        return self._make_new_particle(new_position, new_velocity, old_p.generation + 1)
+
+    def _prepare_data(
+        self, particles: List[Particle]
+    ) -> Tuple[Particle, Particle, Particle]:
+        """
+        Returns the following particles in this very order:
+        1.  old_p: the current particle to be updated now
+        2.  p_best: the personal best value of this particle
+        3.  g_best: the global best value currently known
+        """
+        if len(particles) < self.offspring:
+            raise ValueError("Not enough Particles")
+
+        own_p = [
+            x
+            for x in particles
+            if (isinstance(x, Particle) and x.g_rank == self.rank)
+            or x.rank == self.rank
+        ]
+        if len(own_p) > 0:
+            old_p = max(own_p, key=lambda p: p.generation)
+        else:
+            victim = max(particles, key=lambda p: p.generation)
+            old_p = self._make_new_particle(
+                victim.position, victim.velocity, victim.generation
+            )
```

Comment:
The ``_work`` method does not ensure that any particles are in the list that is passed to the propagator.
The real question is, if I should instead of more or less creating a random particle rather raise a ValueError.